### PR TITLE
Lra 560 lsa ride share new reservation form

### PR DESCRIPTION
**Pull request by britaumich**
- gem simple_calendar is added for calendars (month, week) interfaces (with the stylesheet and simple_calendar templates) 
- gem requestjs-rails is added to handle updating turbo-frame 'available_cars' from reservation_controller.js (availableCars function).
- flatpickr was added to use in the select date field
- ProgramsController's methods get_programs_list, get_students_list and get_sites_list are used by the reservation_controller.js to populate dropdowns menus in the reservation form
- ApplicationHelper methods available_ranges(car, day), available?(car, range), available_time(day), available_cars(cars, range) are used to calculate time ranges when cars are available or check if a car is available for a certain time range.

Note: In the Student model scope :eligible_drivers is a fake one - it should be created in a separate ticket.

The PR has code to create a simple one day reservation for the admin interface (a different ticket for the student interface):
- Unit dropdown is displayed if an admin belongs to more than one unit
- The admin selects a unit (if there is a list) and a term - a program list is populated
- The admin selects a program - a sites list is populated
- The admin selects or changes a day of the reservation - a set of cars cards with available time is populated
- All following changes will update available cars
  - Number of people on the trip
  - Start time and end time
- Finally, the admin selects a car and submits the form
- On the next screen the admin adds a driver and a backup driver
- On the last screen the admin adds passengers to the reservation

There are no validations in this PR - different tickets will be created.
Separate tickets will be created for a reservation Show page, for the editing process, and perhaps something else that I don't know at this moment.

To test:

- login as yourself and test as a super admin
- create a Unit 'Test Unit' with 'test-unit-rideshare-admins' group
- login as um171712 to test a as a single unit admin
- Click on Reservations in the header and then on Add Reservation in a calendar

### CODE REVIEW

File: app/models/student.rb [hunk 1]
```diff
@@ -23,6 +23,8 @@ class Student < ApplicationRecord
 
   validates :uniqname, uniqueness: { scope: :program, message: "is already in the program list" }
 
+  scope :eligible_drivers, -> { where.not(class_training_date: nil) }
```
Comment (rsmoke) [inline app/models/student.rb:26]: Wait until you use a call to this before adding it the codebase

File: app/javascript/controllers/flatpickr_controller.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+import { Application } from '@hotwired/stimulus'
+import Flatpickr from 'stimulus-flatpickr'
```
Comment (rsmoke) [inline app/javascript/controllers/flatpickr_controller.js:2]: The tailwinds component implementation offers this same functionality. https://flowbite.com/docs/plugins/datepicker/

File: app/models/student.rb (hunk 1 above)
Comment (britaumich) [reply app/models/student.rb:26]: I did call it in the reservation_controller in the add_drivers. You can choose drivers only from eligible_drivers. 

```
def add_drivers
    @drivers = @reservation.program.students.eligible_drivers
    unless is_admin?(current_user)
      driver = Student.find_by(program_id: @reservation.program_id, uniqname: current_user.uniqname)
      @reservation.update(driver_id: driver.id)
    end
end
```
I created a ticket to write the code for that scope. I did exactly what you suggested before - create placeholders.

File: app/javascript/controllers/flatpickr_controller.js (hunk 2 above)
Comment (britaumich) [reply app/javascript/controllers/flatpickr_controller.js:2]: But you still need to install flowbite and then add datepicker.js.
Should we explore that and do it in a different ticket? 
It's not essential for this PR how a calendar is displayed in a date field. I did the same thing we did in AVM.

File: app/models/student.rb (hunk 1 above)
Comment (rsmoke) [reply app/models/student.rb:26]: The :eligible_drivers scope is defined in LRA-595

File: app/javascript/controllers/flatpickr_controller.js (hunk 2 above)
Comment (rsmoke) [reply app/javascript/controllers/flatpickr_controller.js:2]: @britaumich I wanted to be sure you were aware of this library

**Review by rsmoke — APPROVED** [review]

**Pull request merged**